src/app/src/test/kotlin/org/vaccineimpact/api/tests/logic/CoverageLogic/SaveCoverageTests.kt

```diff
@@ -0,0 +1,64 @@
+            CoverageIngestionRow("HepB_BD", "AFG", ActivityType.CAMPAIGN, GAVISupportLevel.WITH, 2020, 1, 10, 100F, 78.8F),
+            CoverageIngestionRow("HepB_BD", "AFG", ActivityType.ROUTINE, GAVISupportLevel.WITH, 2022, 1, 10, 100F, 65.5F),
+            CoverageIngestionRow("HepB_BD", "AFG", ActivityType.CAMPAIGN, GAVISupportLevel.WITHOUT, 2023, 1, 10, 100F, 65.5F),
+            CoverageIngestionRow("HepB_BD", "AFG", ActivityType.ROUTINE, GAVISupportLevel.WITHOUT, 2024, 1, 10, 100F, 65.5F),
+            CoverageIngestionRow("HepB", "AFG", ActivityType.CAMPAIGN, GAVISupportLevel.WITH, 2025, 1, 10, 100F, 65.5F),
+            CoverageIngestionRow("HepB", "AFG", ActivityType.ROUTINE, GAVISupportLevel.WITH, 2025, 1, 10, 100F, 65.5F),
+
+            // belongs to same coverage set as the first row
+            CoverageIngestionRow("HepB_BD", "AFG", ActivityType.CAMPAIGN, GAVISupportLevel.WITH, 2021, 1, 10, 100F, 78.8F)
+    )
+
+    @Test
+    fun `new coverage sets are created as needed`()
+    {
+        val mockRepo = mock<TouchstoneRepository>()
+        val sut = RepositoriesCoverageLogic(mock(), mock(), mockRepo, mock())
+        sut.saveCoverageForTouchstone("t1", testSequence)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB_BD", ActivityType.CAMPAIGN, GAVISupportLevel.WITH)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB_BD", ActivityType.ROUTINE, GAVISupportLevel.WITH)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB_BD", ActivityType.CAMPAIGN, GAVISupportLevel.WITHOUT)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB_BD", ActivityType.ROUTINE, GAVISupportLevel.WITHOUT)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB", ActivityType.CAMPAIGN, GAVISupportLevel.WITH)
+        verify(mockRepo, Times(1)).createCoverageSet("t1", "HepB", ActivityType.ROUTINE, GAVISupportLevel.WITH)
+        verify(mockRepo, Times(7)).newCoverageRowRecord(any(), any(), any(), any(), any(), any(), any())
+        verify(mockRepo).saveCoverageForTouchstone(any(), any())
+        verifyNoMoreInteractions(mockRepo)
+    }
+
+
+    @Test
+    fun `coverage records are saved to correct coverage set`()
+    {
+        val mockRepo = mock<TouchstoneRepository> {
+            on { createCoverageSet("t1", "HepB_BD", ActivityType.CAMPAIGN, GAVISupportLevel.WITH) } doReturn 1
+            on { createCoverageSet("t1", "HepB_BD", ActivityType.ROUTINE, GAVISupportLevel.WITH) } doReturn 2
+            on { createCoverageSet("t1", "HepB_BD", ActivityType.CAMPAIGN, GAVISupportLevel.WITHOUT) } doReturn 3
+            on { createCoverageSet("t1", "HepB_BD", ActivityType.ROUTINE, GAVISupportLevel.WITHOUT) } doReturn 4
+        }
+        val sut = RepositoriesCoverageLogic(mock(), mock(), mockRepo, mock())
+        sut.saveCoverageForTouchstone("t1", testSequence)
+        verify(mockRepo, Times(1)).newCoverageRowRecord(eq(1), any(), eq(2020), any(), any(), any(), any())
```

Comment:
Any reason not verify the other parameters?

Reply:
Giving each row that should belong to new coverage set a different year was just a way of verifying that the right rows get added to the correct coverage sets. But you're right that there's no test verifying all params, will add as a separate test so that the purpose of this one remains clear :+1:

Reply:
Adding this extra test found a bug!! Thanks!